---
title: 'Data Preprocessing'
output:
  html_document:
    code_folding: 'hide'
---

```{r load_packages, message=FALSE, warning=FALSE}

library(tidyverse) ; library(reshape2) ; library(glue) ; library(plotly) ; library(plotlyutils)
library(RColorBrewer) ; library(viridis) ; require(gridExtra) ; library(GGally) ; library(ggExtra) ; library(ggpubr)
library(biomaRt) ; library(DESeq2) ; library(sva) ; library(vsn) ; library(WGCNA) ; 
library(GEOquery) ; library(lumi) ; library(limma)
library(dendextend) ; library(expss)
library(knitr) ; library(kableExtra)

```

---

## Raw data
<br><br>

Dataset downloaded from [Arkinglab website](http://www.arkinglab.org/resources/) in the *Transcriptome analysis reveals dysregulation of innate immune response genes and neuronal activity-dependent genes in autism* section.

### Load and annotate data

```{r load_and_annotate_data, message=FALSE, warning=FALSE}

# NCBI biotype annotation
NCBI_biotype = read.csv('./../../../NCBI/Data/gene_biotype_info.csv') %>% 
               dplyr::rename('ensembl_gene_id'=Ensembl_gene_identifier, 'gene_biotype'=type_of_gene, 
                             'hgnc_symbol'=Symbol) %>% 
               mutate(gene_biotype = ifelse(gene_biotype=='protein-coding','protein_coding', gene_biotype))

###############################################################################################################
# DOWNLOAD AND CLEAN GENE EXPRESSION DATA

# Gene expression data downloaded directly from GEO because the one returned with get GEO was already preprocessed
datExpr = lumiR('./../Data/GSE28521_non-normalized_data.txt.gz') %>% as.matrix
rownames_datExpr = rownames(datExpr) 
datExpr = datExpr %>% data.frame %>% mutate_all(function(x) x %>% as.numeric)
rownames(datExpr) = rownames_datExpr

# Download Metadata
GEO_data = getGEO('GSE28521', destdir='./../Data/')[[1]]
#datExpr = exprs(GEO_data) %>% data.frame   # Already filtered and normalised data  
#datGenes = fData(GEO_data)                 # Only includes information for the filtered genes
datMeta = pData(GEO_data) %>% mutate('ID' = geo_accession, 'Brain_lobe' = `tissue (brain region):ch1`, 
                                     'Diagnosis' = factor(ifelse(`disease status:ch1`=='autism', 'ASD','CTL'), 
                                                          levels = c('CTL','ASD')),
                                     'Subject_ID' = substring(title,1,9)) %>%
          dplyr::select(ID, title, Subject_ID, description, Diagnosis, Brain_lobe)
rownames(datMeta) = paste0('X',datMeta$description)


# LABEL GENES WITH ENSEMBL IDS (THEY COME WITH ILLUMINA IDS)
# Get Biomart information
ensembl = useMart('ENSEMBL_MART_ENSEMBL', dataset='hsapiens_gene_ensembl', host='feb2014.archive.ensembl.org') 
getinfo = c('illumina_humanref_8_v3', 'ensembl_gene_id','external_gene_id', 'entrezgene', 'hgnc_symbol',
            'chromosome_name', 'start_position', 'end_position')
datGenes = getBM(attributes = getinfo, filters='illumina_humanref_8_v3', values = rownames(datExpr), 
                 mart = ensembl)
datGenes = datGenes %>% left_join(NCBI_biotype %>% dplyr::select(-hgnc_symbol), by = 'ensembl_gene_id') %>% 
           add_count(illumina_humanref_8_v3) %>% filter(n == 1 | gene_biotype == 'protein_coding') %>%
           distinct(illumina_humanref_8_v3, .keep_all = TRUE) %>% 
           mutate(length = end_position - start_position)

# Match DatExpr and datGenes rows, and datExpr columns and datMeta rows
datExpr = datExpr[rownames(datExpr) %in% datGenes$illumina_humanref_8_v3,]
datGenes = datGenes[match(rownames(datExpr), datGenes$illumina_humanref_8_v3),]
datMeta = datMeta[match(colnames(datExpr),rownames(datMeta)),]


###############################################################################################################


# GO Neuronal annotations: regex 'neuron' in GO functional annotations and label the genes that make a match as neuronal
GO_annotations = read.csv('./../Data/genes_GO_annotations.csv')
GO_neuronal = GO_annotations %>% filter(grepl('neuron', go_term)) %>% 
              mutate('ID'=as.character(ensembl_gene_id)) %>% 
              dplyr::select(-ensembl_gene_id) %>% distinct(ID) %>%
              mutate('Neuronal'=1)


# SFARI Genes
SFARI_genes = read_csv('./../../../SFARI/Data/SFARI_genes_01-03-2020_w_ensembl_IDs.csv')
SFARI_genes = SFARI_genes[!duplicated(SFARI_genes$ID) & !is.na(SFARI_genes$ID),]


# NCBI biotype annotation
NCBI_biotype = read.csv('./../../../NCBI/Data/gene_biotype_info.csv') %>% 
               dplyr::rename('ensembl_gene_id'=Ensembl_gene_identifier, 'gene_biotype'=type_of_gene, 
                             'hgnc_symbol'=Symbol) %>% 
               mutate(gene_biotype = ifelse(gene_biotype=='protein-coding','protein_coding', gene_biotype))


gg_colour_hue = function(n) {
  hues = seq(15, 375, length = n+1)
  pal = hcl(h = hues, l = 65, c = 100)[1:n]
}

rm(GEO_data, GO_annotations, ensembl, getinfo, mart, rownames_datExpr)
```

### Check sample composition
<br><br>

RNA-Seq for 79 cortical brain-tissue samples across frontal, temporal lobes and cerebellum, comprising 29 samples from control subjects and 29 samples from ASD subjects

<br>
The dataset includes `r nrow(datExpr)` genes from `r  ncol(datExpr)` samples belonging to `r length(unique(datMeta$Subject_ID))` different subjects.
<br>

**Counts distribution:** Heavy right tail
```{r, warning=FALSE, message=FALSE}

counts = datExpr %>% melt %>% mutate(value = value %>% as.numeric)

count_distr = data.frame('Statistic' = c('Min', '1st Quartile', 'Median', 'Mean', '3rd Quartile', 'Max'),
                         'Values' = c(min(counts$value), quantile(counts$value, probs = c(.25, .5)) %>% unname,
                                      mean(counts$value), quantile(counts$value, probs = c(.75)) %>% unname,
                                      max(counts$value)))

count_distr %>% kable(digits = 2, format.args = list(scientific = FALSE)) %>% kable_styling(full_width = F)


rm(counts, count_distr)
```
<br>

**Diagnosis distribution by Sample:** Balanced
```{r}
table_info = datMeta %>% apply_labels(Diagnosis = 'Diagnosis', Brain_lobe = 'Brain Lobe')
cro(table_info$Diagnosis)
```
<br>

**Diagnosis distribution by Subject:** Balanced
```{r}
cro(table_info$Diagnosis[!duplicated(table_info$Subject_ID)])
```


**Brain region distribution:**  The Frontal lobe has more samples, but they are quite balanced
```{r}
cro(table_info$Brain_lobe)
```
<br>

Balanced
```{r}
cro(table_info$Diagnosis, list(table_info$Brain_lobe,total()))
```
<br>

**Note:** No age or gender information in the metadata :/

<br><br>

### Annotate genes with BioMart information
<br><br>

I was originally running this with the feb2014 version of BioMart because that's the one that Gandal used (and it finds all of the Ensembl IDs, which other versions don't), but it has some outdated biotype annotations, to fix this I'll obtain all the information except the biotype label from BioMart in the same way as it had been done before, and then I'll add the most current biotype label using information from NCBI's website and then from BioMart in the following way:

1. Use BioMart to run a query with the original feb2014 version using the Ensembl IDs as keys to obtain all the information except the biotype labels (we are not going to retrieve the gene name from biomart because we already extracted it from datExpr)

2. Annotate genes with Biotype labels:

    2.1 Use the NCBI annotations downloaded from [NCBI's website](https://www.ncbi.nlm.nih.gov/home/download/) and processed in NCBI/RMarkdowns/clean_data.html (there is information for only 26K genes, so some genes will remain unlabelled)
  
    2.2 Use the current version (jan2020) to obtain the biotype annotations using the Ensembl ID as keys (some genes don't return a match)
  
    2.3 For the genes that didn't return a match, use the current version (jan2020) to obtain the biotype annotations using the gene name as keys
  
    2.4 For the genes that returned multiple labels, use the feb2014 version with the Ensembl IDs as keys

<br>

**Note:** A small proportion of genes don't make a match in any of these queries, so they will be lost when we start filtering out genes
<br>
```{r annotate_genes, warning=FALSE, message=FALSE, fig.height=2, fig.align='center'}

labels_source = data.frame('source' = c('NCBI', 'BioMart2020_byID', 'BioMart2020_byGene', 'BioMart2014'),
                                      'n_matches' = rep(0,4))

########################################################################################
# 1. Query archive version

# We already did this part when loading the data

cat(paste0('1. ', sum(is.na(datGenes$start_position)), '/', nrow(datGenes),
             ' Ensembl IDs weren\'t found in the feb2014 version of BioMart'))


########################################################################################
########################################################################################
# 2. Get Biotype Labels

cat('2. Add biotype information')

########################################################################################
# 2.1 Add NCBI annotations

# Already did this as well

cat(paste0('2.1 ' , sum(is.na(datGenes$gene_biotype)), '/', nrow(datGenes),
             ' Ensembl IDs weren\'t found in the NCBI database'))

labels_source$n_matches[1] = sum(!is.na(datGenes$gene_biotype))

########################################################################################
# 2.2 Query current BioMart version for gene_biotype using Ensembl ID as key

getinfo = c('ensembl_gene_id','gene_biotype')
mart = useMart(biomart = 'ENSEMBL_MART_ENSEMBL', dataset = 'hsapiens_gene_ensembl',
               host = 'jan2020.archive.ensembl.org')
datGenes_biotype = getBM(attributes = getinfo, filters = c('ensembl_gene_id'), mart=mart, 
                         values = datGenes$ensembl_gene_id[is.na(datGenes$gene_biotype)])

cat(paste0('2.2 ' , sum(is.na(datGenes$gene_biotype))-nrow(datGenes_biotype), '/', 
           sum(is.na(datGenes$gene_biotype)),
           ' Ensembl IDs weren\'t found in the jan2020 version of BioMart when querying by Ensembl ID'))

# Add new gene_biotype info to datGenes
datGenes = datGenes %>% left_join(datGenes_biotype, by='ensembl_gene_id') %>%
           mutate(gene_biotype = coalesce(as.character(gene_biotype.x), gene_biotype.y)) %>%
           dplyr::select(-gene_biotype.x, -gene_biotype.y)

labels_source$n_matches[2] = sum(!is.na(datGenes$gene_biotype)) - labels_source$n_matches[1]

########################################################################################
# 3. Query current BioMart version for gene_biotype using gene symbol as key

missing_genes = unique(datGenes$external_gene_id[is.na(datGenes$gene_biotype)])
getinfo = c('hgnc_symbol','gene_biotype')
datGenes_biotype_by_gene = getBM(attributes = getinfo, filters = c('hgnc_symbol'), mart = mart,
                                 values = missing_genes)

cat(paste0('2.3 ', length(missing_genes)-length(unique(datGenes_biotype_by_gene$hgnc_symbol)),'/',
           length(missing_genes),
           ' genes weren\'t found in the current BioMart version when querying by gene name'))

dups = unique(datGenes_biotype_by_gene$hgnc_symbol[duplicated(datGenes_biotype_by_gene$hgnc_symbol)])
cat(paste0('    ', length(dups), ' genes returned multiple labels (these won\'t be added)'))

# Update information
datGenes_biotype_by_gene = datGenes_biotype_by_gene %>% filter(!hgnc_symbol %in% dups)
datGenes = datGenes %>% left_join(datGenes_biotype_by_gene, by=c('external_gene_id'='hgnc_symbol')) %>% 
           mutate(gene_biotype = coalesce(gene_biotype.x, gene_biotype.y)) %>%
           dplyr::select(-gene_biotype.x, -gene_biotype.y) %>%
           mutate(hgnc_symbol = ifelse(is.na(hgnc_symbol), external_gene_id, hgnc_symbol))

labels_source$n_matches[3] = sum(!is.na(datGenes$gene_biotype)) - sum(labels_source$n_matches)

########################################################################################
# 4. Query feb2014 BioMart version for the missing biotypes

missing_ensembl_ids = unique(datGenes$ensembl_gene_id[is.na(datGenes$gene_biotype)])

getinfo = c('ensembl_gene_id','gene_biotype')
mart = useMart(biomart = 'ENSEMBL_MART_ENSEMBL', dataset = 'hsapiens_gene_ensembl', 
               host = 'feb2014.archive.ensembl.org')
datGenes_biotype_archive = getBM(attributes = getinfo, filters=c('ensembl_gene_id'), 
                                 values = missing_ensembl_ids, mart=mart)

cat(paste0('2.4 ', length(missing_ensembl_ids)-nrow(datGenes_biotype_archive),'/',length(missing_ensembl_ids),
             ' genes weren\'t found in the feb2014 BioMart version when querying by Ensembl ID'))

# Update information
datGenes = datGenes %>% left_join(datGenes_biotype_archive, by='ensembl_gene_id') %>% 
            mutate(gene_biotype = coalesce(gene_biotype.x, gene_biotype.y)) %>%
            dplyr::select(-gene_biotype.x, -gene_biotype.y)

labels_source$n_matches[4] = sum(!is.na(datGenes$gene_biotype)) - sum(labels_source$n_matches)

########################################################################################
# Plot results

labels_source = labels_source %>% add_row(source = 'missing', 
                                          n_matches = nrow(datGenes) - sum(labels_source$n_matches)) %>% 
                mutate(x = 1, percentage = round(100*n_matches/sum(n_matches),2),
                       source = factor(source, levels=c('BioMart2014','BioMart2020_byGene','BioMart2020_byID',
                                                        'NCBI','missing')))
                

p = labels_source %>% ggplot(aes(x, percentage, fill=source)) + geom_bar(position='stack', stat='identity') +
    theme_minimal() + coord_flip() + theme(legend.position='bottom', axis.title.y=element_blank(),
    axis.text.y=element_blank(), axis.ticks.y=element_blank()) + ylab('Percentage of genes') +
    scale_fill_manual(values = c(gg_colour_hue(nrow(labels_source)-1),'gray'))

ggplotly(p + theme(legend.position='none'))
as_ggplot(get_legend(p))

########################################################################################
# Reorder rows to match datExpr
datGenes = datGenes[match(rownames(datExpr), datGenes$illumina_humanref_8_v3),]


rm(getinfo, mart, datGenes_BM, datGenes_biotype, datGenes_biotype_by_gene, datGenes_biotype_archive,
   dups, missing_ensembl_ids, missing_genes, labels_source, p)
```

---

## Filtering
<br><br>

Checking how many SFARI genes are in the dataset

```{r}
df = SFARI_genes %>% dplyr::select(-gene_biotype) %>% inner_join(datGenes, by=c('ID'='ensembl_gene_id'))
n_SFARI = df[['gene-symbol']] %>% unique %>% length
```

Considering all genes, this dataset contains `r df[['gene-symbol']] %>% unique %>% length` of the `r SFARI_genes[['gene-symbol']] %>% unique %>% length` SFARI genes
<br><br>

**1.-** Filter entries for which we didn't manage to obtain its genotype information
<br>

1.1 Missing Biotype
```{r filter_genes_wo_length}

to_keep = !is.na(datGenes$gene_biotype)

datGenes = datGenes[to_keep,]
datExpr = datExpr[to_keep,]

```

Removed `r sum(!to_keep)` 'genes', `r sum(to_keep)` remaining
<br>

Filtering genes without biotype information, we are left with `r df[['gene-symbol']][!is.na(df$gene_biotype)] %>% unique %>% length` SFARI Genes (we lose `r n_SFARI - df[['gene-symbol']][!is.na(df$gene_biotype)] %>% unique %>% length` genes)

<br><br><br>


**2.-** Filter genes that do not encode any protein

<br>

`r round(100*mean(datGenes$gene_biotype == 'protein_coding'))`% of the genes are protein coding genes

```{r gene_biotype_table}
datGenes$gene_biotype %>% table %>% sort(decreasing=TRUE) %>% kable(caption='Biotypes of genes in dataset') %>%
                          kable_styling(full_width = F)
```

Non-protein coding genes in general have lower levels of expression than protein coding genes, but the difference is not that big
```{r explore_non_protein_coding_genes, fig.width=10}

plot_data = data.frame('ID' = rownames(datExpr), 'MeanExpr' = rowMeans(datExpr),
                       'ProteinCoding' = datGenes$gene_biotype=='protein_coding')

ggplotly(plot_data %>% ggplot(aes(log2(MeanExpr+1), fill=ProteinCoding, color=ProteinCoding)) + 
         geom_density(alpha=0.5) + theme_minimal())


rm(plot_data)
```


```{r protein_coding_genes_and_SFARI_score}

df = SFARI_genes %>% dplyr::select(-gene_biotype) %>% inner_join(datGenes, by=c('ID'='ensembl_gene_id'))

```

Filtering protein coding genes, we are left with `r df[['gene-symbol']][df$gene_biotype=='protein_coding'] %>% unique %>% length` SFARI Genes (we lose `r n_SFARI - df[['gene-symbol']][df$gene_biotype=='protein_coding'] %>% unique %>% length` genes)
<br>


```{r filter_non_protein_coding_genes}

if(!all(rownames(datExpr)==rownames(datGenes))) cat('!!! gene rownames do not match!!!')

to_keep = datGenes$gene_biotype == 'protein_coding'
datExpr = datExpr %>% filter(to_keep)
datGenes = datGenes %>% filter(to_keep)

```
<br>

Removed `r sum(!to_keep)` genes. `r sum(to_keep)` remaining


<br>
**3.-** Filter genes with low expression levels

<br>

Choosing the threshold:

**Criteria for selecting filtering threshold:** The minimum value in which the preprocessed data is relatively homoscedastic (we're trying to get rid of the group of genes with very low mean and SD that make the cloud of points look like a comic book speech bubble)

- The Mean vs SD plot doesn't show all of the genes, a random sample was selected for the genes with higher level of expression so the visualisation wouldn't be as heavy (and since we care about the genes with the lowest levels of expression, we aren't losing important information)

```{r}
datMeta_original = datMeta
datExpr_original = datExpr
datGenes_original = datGenes
```

```{r warning=FALSE, message=FALSE, results='hide'}

thresholds = c(100, 200, 220, 250, 270, 300)

for(threshold in thresholds){
  
  datMeta = datMeta_original
  datExpr = datExpr_original
  datGenes = datGenes_original
  
  to_keep = apply(datExpr, 1, function(x) mean(x) >= threshold)
  datGenes = datGenes[to_keep,]
  datExpr = datExpr[to_keep,]
  
  # Filter outlier samples
  absadj = datExpr %>% bicor %>% abs
  netsummary = fundamentalNetworkConcepts(absadj)
  ku = netsummary$Connectivity
  z.ku = (ku-mean(ku))/sqrt(var(ku))
  
  to_keep = z.ku > -2
  datMeta = datMeta[to_keep,]
  datExpr = datExpr[,to_keep]
  
  
  # Normaise data using variance stabilisation normalisation
  LumiBatch = ExpressionSet(assayData = datExpr %>% as.matrix)
  pData(LumiBatch) = datMeta
  LumiBatch = lumiN(LumiBatch, method = 'vsn', verbose = FALSE)
  datExpr = exprs(LumiBatch)
  
  rm(absadj, netsummary, ku, z.ku, to_keep, LumiBatch)
  
  
  # Save summary results in dataframe
  if(threshold == thresholds[1]){
    mean_vs_sd_data = data.frame('threshold' = threshold, 'ID' = rownames(datExpr),
                                 'Mean' = rowMeans(datExpr), 'SD' = apply(datExpr,1,sd))
  } else {
    new_entries = data.frame('threshold' = threshold, 'ID' = rownames(datExpr),
                                 'Mean' = rowMeans(datExpr), 'SD' = apply(datExpr,1,sd))
    mean_vs_sd_data = rbind(mean_vs_sd_data, new_entries)
  }
}


# Visualise the effects of different thresholds
to_keep_1 = mean_vs_sd_data$ID[mean_vs_sd_data$threshold==thresholds[1] & mean_vs_sd_data$Mean<7] %>%
            as.character
to_keep_2 = mean_vs_sd_data$ID[mean_vs_sd_data$threshold==thresholds[1] & mean_vs_sd_data$Mean>=7]
to_keep_2 = to_keep_2 %>% sample(round(length(to_keep_2)/10)) %>% as.character

```

```{r fig.width=10, warning=FALSE, message=FALSE}

plot_data = mean_vs_sd_data[mean_vs_sd_data$ID %in% c(to_keep_1, to_keep_2),]

ggplotly(plot_data %>% ggplot(aes(Mean, SD)) + 
         geom_point(color='#0099cc', alpha=0.2, aes(id=ID, frame=threshold)) + 
         scale_x_log10() + scale_y_log10() + theme_minimal())

# Plot remaining genes
plot_data = mean_vs_sd_data %>% group_by(threshold) %>% tally

ggplotly(plot_data %>% ggplot(aes(threshold, n)) + geom_point() + geom_line() +
         theme_minimal() + ggtitle('Remaining genes for each filtering threshold'))


rm(to_keep_1, to_keep_2, plot_data, dds, thresholds)
```

```{r}

# Return to original variables
datExpr = datExpr_original
datGenes = datGenes_original
datMeta = datMeta_original


rm(datExpr_original, datGenes_original, datMeta_original)
```

<br>

Selecting a threshold of 220
```{r, warning=FALSE, fig.width=10}

# Minimum percentage of non-zero entries allowed per gene
threshold = 220

plot_data = data.frame('id'=rownames(datExpr),
                       'threshold' = apply(datExpr, 1, function(x) mean(x)))

ggplotly(plot_data %>% ggplot(aes(x=threshold)) + 
         geom_density(color='#0099cc', fill='#0099cc', alpha=0.3) + 
         geom_vline(xintercept=threshold, color='gray') + 
         xlab('Mean level of expression') + ylab('Density') + scale_x_log10() +
         ggtitle('Mean level of expression by Gene') + theme_minimal())

to_keep = apply(datExpr, 1, function(x) mean(x) >= threshold)
datGenes = datGenes[to_keep,]
datExpr = datExpr[to_keep,]

```

Removed `r sum(!to_keep)` genes. `r sum(to_keep)` remaining
<br>

`r SFARI_genes[['gene-symbol']][SFARI_genes$ID %in% datGenes$ensembl_gene_id] %>% unique %>% length` SFARI genes remaining (we lost `r n_SFARI - SFARI_genes[['gene-symbol']][SFARI_genes$ID %in% datGenes$ensembl_gene_id] %>% unique %>% length` genes)

```{r}
n_SFARI = SFARI_genes[['gene-symbol']][SFARI_genes$ID %in%datGenes$ensembl_gene_id] %>% unique %>% length

rm(threshold, plot_data, to_keep)
```

<br>
**4.-** Filter outlier samples

Using node connectivity as a distance measure, normalising it and filtering out genes farther away than 2 standard deviations from the left (lower connectivity than average, not higher)

- Gandal uses the formula $s_{ij}=\frac{1+bw(i,j)}{2}$ to convert all the weights to positive values, but I used $s_{ij}=|bw(i,j)|$ instead because I think it makes more sense. In the end it doesn't matter because they select as outliers the same six samples

- 5 outliers

```{r, warning=FALSE}

absadj = datExpr %>% bicor %>% abs
netsummary = fundamentalNetworkConcepts(absadj)
ku = netsummary$Connectivity
z.ku = (ku-mean(ku))/sqrt(var(ku))

plot_data = data.frame('sample'=1:length(z.ku), 'distance'=z.ku, 'Sample_ID'=datMeta$ID, 
                       'Subject_ID'=datMeta$Subject_ID, 'Brain_Lobe'=datMeta$Brain_lobe, 
                       'Diagnosis'=datMeta$Diagnosis)

selectable_scatter_plot(plot_data, plot_data[,-c(1:3)])

```

Outlier samples: `r paste(as.character(plot_data$Sample_ID[plot_data$distance< -2]), collapse=', ')`

```{r}
to_keep = z.ku > -2
datMeta = datMeta[to_keep,]
datExpr = datExpr[,to_keep]

rm(absadj, netsummary, ku, z.ku, plot_data)
```

Removed `r sum(!to_keep)` samples, `r sum(to_keep)` remaining

```{r}
rm(to_keep)
```
<br><br><br>

<br>
**5.-** Filter repeated genes

There are `r nrow(datGenes) - datGenes$hgnc_symbol %>% unique %>% length` genes with more than one ensembl ID in the dataset. To accurately refer to the rows of my data as 'genes', I'm going to remove the repeated ones.

```{r}
dup_genes = datGenes$hgnc_symbol %>% duplicated

datGenes = datGenes[!dup_genes,]
datExpr = datExpr[!dup_genes,]
```

Removed `r sum(dup_genes)` genes. `r sum(!dup_genes)` remaining
<br>

`r SFARI_genes[['gene-symbol']][SFARI_genes$ID %in% datGenes$ensembl_gene_id] %>% unique %>% length` SFARI genes remaining (we lost `r n_SFARI - SFARI_genes[['gene-symbol']][SFARI_genes$ID %in% datGenes$ensembl_gene_id] %>% unique %>% length` genes)

```{r}
rm(dup_genes, n_SFARI)
```
<br><br>

After filtering, the dataset consists of `r nrow(datExpr)` genes and `r ncol(datExpr)` samples
<br><br>

### Save filtered and annotated dataset

```{r}
save(datExpr, datMeta, datGenes, file='./../Data/filtered_raw_data.RData')
#load('./../Data/filtered_raw_data.RData')
```

---

## Batch Effects Exploratory Analysis
<br><br>

There are no batch surrogate variables in this dataset
<br>

Cerebellum samples seem to cluster together. There doesn't seem to be a strong pattern related to Diagnosis

```{r samples_histogram, fig.width=10}

h_clusts = datExpr %>% t %>% dist %>% hclust %>% as.dendrogram

dend_meta = datMeta[match(labels(h_clusts), rownames(datMeta)),] %>% 
            mutate('Diagnosis' = ifelse(Diagnosis=='CTL','#008080','#86b300'),  # Blue control, Green ASD
                   'Region' = case_when(Brain_lobe=='Frontal cortex'~'#F8766D', # ggplot defaults for 3 colours
                                        Brain_lobe=='Temporal cortex'~'#00BA38',
                                        TRUE~'#619CFF')) %>%
            dplyr::select(Region, Diagnosis)
h_clusts %>% dendextend::set('labels', rep('', nrow(datMeta))) %>% 
             dendextend::set('branches_k_color', k=9) %>% plot
colored_bars(colors=dend_meta)


rm(h_clusts, dend_meta)
```

<br><br>

### Looking for unknown sources of batch effects
<br>

Following the pipeline from [Surrogate variable analysis: hidden batch effects](https://biodatascience.github.io/compbio/dist/sva.html) where sva is used with DESeq2:
<br>

Create a lumi object
```{r, results = 'hide', warning=FALSE, message=FALSE}
# Normaise data using variance stabilisation normalisation
LumiBatch = ExpressionSet(assayData = datExpr %>% as.matrix)
pData(LumiBatch) = datMeta
LumiBatch = lumiN(LumiBatch, method = 'vsn', verbose = FALSE)
datExpr_norm = exprs(LumiBatch)
```

Provide the normalized counts and two model matrices to SVA. The first matrix uses the biological condition, and the second model matrix is the null model.
```{r}

# Perform vst
mod = model.matrix(~ Diagnosis, datMeta)
mod0 = model.matrix(~ 1, datMeta)
sva_fit = sva(datExpr_norm, mod=mod, mod0=mod0)

  
rm(mod, mod0)
```

Found `r ncol(sva_fit$sv)` surrogate variables, since there is no direct way to select which ones to pick [Bioconductor answer](https://support.bioconductor.org/p/71447/), kept all of them.

Include SV estimations to datMeta information
```{r}
sv_data = sva_fit$sv %>% data.frame
colnames(sv_data) = paste0('SV', 1:ncol(sv_data))

datMeta_sva = cbind(datMeta, sv_data)

rm(sv_data, sva_fit)
```
<br><br>

---

## Normalisation and Differential Expression Analysis
<br><br>

Using the lumi package to perform normalisation

```{r}
plot_data = data.frame('ID'=rownames(datExpr), 'Mean'=rowMeans(datExpr), 'SD'=apply(datExpr,1,sd))

plot_data %>% ggplot(aes(Mean, SD)) + geom_point(color='#0099cc', alpha=0.1) + geom_abline(color='gray') +
              scale_x_log10() + scale_y_log10() + theme_minimal()

rm(plot_data)
```

Using vsn to normalise the data

```{r normalisation, warning=FALSE, message=FALSE}

# Normaise data using variance stabilisation normalisation
LumiBatch = ExpressionSet(assayData = datExpr %>% as.matrix)
pData(LumiBatch) = datMeta
LumiBatch = lumiN(LumiBatch, method = 'vsn', verbose = FALSE)

datExpr_vst = exprs(LumiBatch)

# DEA
fit = lmFit(datExpr_vst, design=model.matrix(~Diagnosis, data = datMeta))
efit = eBayes(fit)
DE_info = topTable(efit,coef=2, number=Inf, sort.by='none')

rm(LumiBatch, fit)
```

Using the plotting function [DESEq2's manual](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) proposes to study vst's output it looks like the data could be homoscedastic
```{r}
meanSdPlot(datExpr_vst, plot=FALSE)$gg + theme_minimal()
```

Plotting points individually we can notice some heteroscedasticity in the data in the genes with the lowest levels of expression

```{r warning=FALSE, message=FALSE}

plot_data = data.frame('ID'=rownames(datExpr_vst), 'Mean'=rowMeans(datExpr_vst), 'SD'=apply(datExpr_vst,1,sd))

plot_data %>% ggplot(aes(Mean, SD)) + geom_point(color='#0099cc', alpha=0.2) + geom_smooth(color = 'gray') +
              scale_x_log10() + scale_y_log10() + theme_minimal()

rm(plot_data)
```
<br><br>

Rename normalised datasets to continue working with these
```{r}
datExpr = datExpr_vst
datMeta = datMeta_sva %>% data.frame
#datGenes = datGenes_vst

rm(datExpr_vst, datMeta_vst, datMeta_sva)
```

<br><br>

---

## Batch Effect Correction
<br><br>

### SVA surrogate variables
<br>

In some places they say you shouldn't correct these effects on the data because you risk losing biological variation, in others they say you should because they introduce noise to the data. The only thing everyone agrees on is that you shouldn't remove them before performing DEA but instead include them in the model.

Based on the conclusions from [Practical impacts of genomic data “cleaning” on biological discovery using surrogate variable analysis](https://bmcbioinformatics.biomedcentral.com/track/pdf/10.1186/s12859-015-0808-5) it seems like it may be a good idea to remove the batch effects from the data and not only from the DE analysis:

- Using SVA, ComBat or related tools can increase the power to identify specific signals in complex genomic datasets (they found "greatly sharpened global and gene-specific differential expression across treatment groups")

- But caution should be exercised to avoid removing biological signal of interest

- We must be precise and deliberate in the design and analysis of experiments and the resulting data, and also mindful of the limitations we impose with our own perspective

- Open data exploration is not possible after such supervised “cleaning”, because effects beyond those stipulated by the researcher may have been removed
<br>

#### Comparing data with and without surrogate variable correction

```{r correct_svs_in_datExpr, warning=FALSE}

# Taken from https://www.biostars.org/p/121489/#121500
correctDatExpr = function(datExpr, mod, svs) {
  X = cbind(mod, svs)
  Hat = solve(t(X) %*% X) %*% t(X)
  beta = (Hat %*% t(datExpr))
  rm(Hat)
  gc()
  P = ncol(mod)
  return(datExpr - t(as.matrix(X[,-c(1:P)]) %*% beta[-c(1:P),]))
}

pca_samples_before = datExpr %>% t %>% prcomp
pca_genes_before = datExpr %>% prcomp

# Correct
mod = model.matrix(~ Diagnosis, datMeta)
svs = datMeta %>% dplyr::select(contains('SV')) %>% as.matrix
datExpr_corrected = correctDatExpr(as.matrix(datExpr), mod, svs)

pca_samples_after = datExpr_corrected %>% t %>% prcomp
pca_genes_after = datExpr_corrected %>% prcomp


rm(correctDatExpr)
```

#### Samples
<br>

The data is divided into two very different groups but the SVA manages to join them together and make Diagnosis the primary characteristic that separetes them
```{r pca_samples, warning=FALSE}

pca_samples_df = rbind(data.frame('ID'=colnames(datExpr), 'PC1'=pca_samples_before$x[,1],
                                  'PC2'=pca_samples_before$x[,2], 'corrected'=0),
                       data.frame('ID'=colnames(datExpr), 'PC1'=pca_samples_after$x[,1],
                                  'PC2'=pca_samples_after$x[,2], 'corrected'=1)) %>%
                 left_join(datMeta %>% mutate('ID'=rownames(datMeta)), by='ID')

ggplotly(pca_samples_df %>% ggplot(aes(PC1, PC2, color=Diagnosis)) + geom_point(aes(frame=corrected, id=ID), alpha=0.75) + 
         xlab(paste0('PC1 (corr=', round(cor(pca_samples_before$x[,1],pca_samples_after$x[,1]),2),
                     '). % Var explained: ', round(100*summary(pca_samples_before)$importance[2,1],1),' to ',
                     round(100*summary(pca_samples_after)$importance[2,1],1))) +
         ylab(paste0('PC2 (corr=', round(cor(pca_samples_before$x[,2],pca_samples_after$x[,2]),2),
                     '). % Var explained: ', round(100*summary(pca_samples_before)$importance[2,2],1),' to ',
                     round(100*summary(pca_samples_after)$importance[2,2],1))) +
         ggtitle('Samples') + theme_minimal())

```
<br>

The group of samples that has a very different behaviour to the rest of the genes at the beginning are the ones from the Cerebellum. After performing SVA, they are no longer recognisable
```{r pca_samples_brain_region, warning=FALSE}

ggplotly(pca_samples_df %>% ggplot(aes(PC1, PC2, color=Brain_lobe)) + geom_point(aes(frame=corrected, id=ID), alpha=0.75) + 
         xlab(paste0('PC1 (corr=', round(cor(pca_samples_before$x[,1],pca_samples_after$x[,1]),2),
                     '). % Var explained: ', round(100*summary(pca_samples_before)$importance[2,1],1),' to ',
                     round(100*summary(pca_samples_after)$importance[2,1],1))) +
         ylab(paste0('PC2 (corr=', round(cor(pca_samples_before$x[,2],pca_samples_after$x[,2]),2),
                     '). % Var explained: ', round(100*summary(pca_samples_before)$importance[2,2],1),' to ',
                     round(100*summary(pca_samples_after)$importance[2,2],1))) +
         ggtitle('Samples') + theme_minimal())

rm(pca_samples_df)
```

#### Genes
<br>

It seems like the sva correction preserves the mean expression of the genes and erases almost everything else (although what little else remains is enough to characterise the two Diagnosis groups pretty well using only the first PC)

*Plot is done with only 10% of the genes so it's not that heavy
```{r pca_genes, warning=FALSE, message=FALSE}
pca_genes_df = rbind(data.frame('ID'=rownames(datExpr), 'PC1'=pca_genes_before$x[,1],
                                'PC2'=pca_genes_before$x[,2], 'corrected'=0, 'MeanExpr'=rowMeans(datExpr)),
                     data.frame('ID'=rownames(datExpr), 'PC1'=pca_genes_after$x[,1],
                                'PC2'=pca_genes_after$x[,2], 'corrected'=1, 'MeanExpr'=rowMeans(datExpr)))

keep_genes = rownames(datExpr) %>% sample(0.1*nrow(datExpr))

pca_genes_df = pca_genes_df %>% filter(ID %in% keep_genes)

ggplotly(pca_genes_df %>% ggplot(aes(PC1, PC2,color=MeanExpr)) + 
         geom_point(alpha=0.3, aes(frame=corrected, id=ID)) +
         xlab(paste0('PC1 (corr=', round(cor(pca_genes_before$x[,1],pca_genes_after$x[,1]),2),
                     '). % Var explained: ', round(100*summary(pca_genes_before)$importance[2,1],1),' to ',
                     round(100*summary(pca_genes_after)$importance[2,1],1))) +
         ylab(paste0('PC2 (corr=', round(cor(pca_genes_before$x[,2],pca_genes_after$x[,2]),2),
                     '). % Var explained: ', round(100*summary(pca_genes_before)$importance[2,2],1),' to ',
                     round(100*summary(pca_genes_after)$importance[2,2],1))) +
         scale_color_viridis() + ggtitle('Genes') + theme_minimal())


rm(pca_samples_before, pca_genes_before, mod, svs, pca_samples_after, pca_genes_after, pca_genes_df, keep_genes)
```


Everything looks good, so we're keeping the corrected expression dataset
```{r}

datExpr = datExpr_corrected


rm(datExpr_corrected)
```
<br>

---

#### Save preprocessed dataset
```{r save_preprocessed_dataset}

save(datExpr, datMeta, datGenes, DE_info, efit, file='./../Data/preprocessed_data.RData')

```
<br><br>

---

#### Session info
```{r print_session_info}

sessionInfo()

```
<br><br>
